### PR TITLE
Rover: HOLD mode in AUTO will keep reversed flag

### PR DESCRIPTION
**Pull request by robot-to-society**
Currently, changing the hold mode from AUTO resets the reverse flag, which is not the correct behavior for a mowing mission that contains a DO_SET_REVERSE command. I would like to be able to maintain the reverse flag while the mower goes from AUTO -> HOLD -> AUTO.

This is a video tested in SITL.
https://drive.google.com/file/d/1QhU1QoM_vY50oIVMzYeWu1zqfvpIl-3N/view?usp=sharing

### CODE REVIEW

File: Rover/system.cpp [hunk 1]
```diff
@@ -219,6 +219,18 @@ bool Rover::set_mode(Mode &new_mode, ModeReason reason)
     }
 
     Mode &old_mode = *control_mode;
+
+    // HOLD mode in AUTO will keep reversed flag
+    if (!((old_mode.mode_number() == Mode::Number::AUTO && new_mode.mode_number() == Mode::Number::HOLD) ||
+        (old_mode.mode_number() == Mode::Number::HOLD && new_mode.mode_number() == Mode::Number::AUTO))) {
```
Comment (IamPete1) [inline Rover/system.cpp:225]:
```suggestion
    if (!((old_mode == &mode_auto && new_mode == &mode_hold) ||
        (&mode_hold== Mode::Number::HOLD && new_mode == &mode_auto))) {
```
Pointer comparison is better.

File: Rover/system.cpp [hunk 2]
```diff
@@ -219,6 +219,18 @@ bool Rover::set_mode(Mode &new_mode, ModeReason reason)
...
+        new_mode.set_reversed(false);
+    }
+    // Restart and completetion should reset reversed flag
+    if ((new_mode.mode_number() == Mode::Number::AUTO && !rover.mode_auto.mission.is_resume()) ||
```
Comment (IamPete1) [inline Rover/system.cpp:229]:
```suggestion
    if ((new_mode == &mode_auto) && !rover.mode_auto.mission.is_resume()) ||
```

**Review by IamPete1 — COMMENTED** [review]
I wonder if we could make it work more generally. We have the `is_autopilot_mode` method. Maybe we only reset if entering a autopilot mode which is not auto, and then we add a reset on mission start (IE not resume).

**Comment by robot-to-society** [comment]
@IamPete1 Thanks for your comment. I agree with your suggestion. But I think it should be reset in non-autopilot modes other than HOLD as well. If you switch from AUTO to MANUAL mode while in reverse, does that result in the user's forward control going in reverse?

File: libraries/AP_Mission/AP_Mission.h [hunk 3]
```diff
@@ -763,6 +763,8 @@ class AP_Mission
 
     bool is_valid_index(const uint16_t index) const { return index < _cmd_total; }
 
+    bool is_resume() const { return _restart == 0; }
```
Comment (rmackay9) [inline libraries/AP_Mission/AP_Mission.h:766]: Let's add a comment above this new method, "returns true if configured to resume"

I also think we should move this up to be closer to other related methods.  I think directly below "set_force_resume()" might be good.

Also it makes no difference to Rover I think the return should also check "_force_resume"

Comment (robot-to-society) [reply libraries/AP_Mission/AP_Mission.h]: @rmackay9 Thank you for your comment. I fixed it.

**Comment by rmackay9** [comment]
Hi @robot-to-society,  looking good!   Can you squash all the commits together?

**Comment by robot-to-society** [comment]
@rmackay9 Thank you for review this PR. It squashed and change commit message a little bit. Is this OK?

**Comment by rmackay9** [comment]
Hi @robot-to-society,

Ah, it looks like the 1st commit changes both AP_Mission and Rover.  There should be two commits, one that affects AP_Mission and the other Rover.  I think you can handle git but if you want me to fix it up I can.

Thanks!

**Comment by robot-to-society** [comment]
@rmackay9 I think it's fixed now.

**Review by rmackay9 — APPROVED** [review]
looks good, thanks!

File: Rover/mode.cpp [hunk 4]
```diff
@@ -44,7 +44,10 @@ bool Mode::enter()
 
     // initialisation common to all modes
     if (ret) {
-        set_reversed(false);
+        // reset reversed for autopilot mode except auto
+        if (is_autopilot_mode() && mode_number() != Mode::Number::AUTO) {
```
Comment (peterbarker) [inline Rover/mode.cpp:48]:
```suggestion
        init_reversed_flag()
```

... and crate a virtual method on the Mode object which checks `is_autopilot_mode`, and overridden by ModeAuto to check the is_resume flag.

Comment (robot-to-society) [reply Rover/mode.cpp]: @peterbarker Thank you for your comment. I'm not sure if I understood your advice properly, but I fixed it.

**Review by peterbarker — APPROVED** [review]
LGTM, thanks!

**Pull request merged**